### ISSUE TITLE
Not an issue but a question...

### ISSUE THREAD

**Issue by born2net** [InfomediaLtd/angular2-redux-example#1]
I am debating between Redux and the Observable version of it https://github.com/ngrx/store
I am leaning towards the https://github.com/ngrx/store as I love the fact it uses pure Rxjs and we can pipe async into the view... 

but, I like the middleware and eco system aroundd Redux... 

decisions decisions..... mmm

u'r 2 cents?

Regards,

Sean

Checkout the Ultimate Angular 2 Boorstrap App: @ http://ng2.javascriptninja.io
Source@ https://github.com/born2net/ng2Boilerplate

**Comment by rubyboy** [comment]
@born2net That's a great question. I've started with the pure redux implementation before ngrx was implemented by Rob and I have to say I'm still quite happy with the decision (not to take away any of the awesomeness of his work!).

A bit of background: At our company, we are going to start a band new project with Angular 2 in the next month or two. There are going to be 20-30 developers initially and we need something solid for them to work on. Angular 2 + TypeScript are good foundations, but they really needed an opinionated way to handle state and logic to compliment the architecture. When I first read about Redux and then went through the learning material I thought it was perfect. Fits nicely with the way we are planning to build the application and provides a beautiful separation between state, logic and UI. Also, it's very consistent and predictable in the way developers need to write and maintain code (As an aside: the only dilemma I have is around thunks and the way they are different than other action creators by having side effects). Creating this angular2-redux-example proved that this is a viable option that would work for a bigger application and I could see our developers easily adopting it and understanding how it works. It's just so simple that it hurts! :)

Ngrx looks very similar to use, but I'm afraid it's not as simple to grok when looking at the internals. Rx in general is not a simple concept to grasp. I love it and think this is an awesome way to handle data streams and events, but it will take a while until it's mainstream and developers just "get it". I don't think I fully get it yet and think this way. So, I'm very careful in adopting it. Angular 2, having RxJS build it, will help get a head start with RxJS when dealing with forms or the Http service, so once we do use it for a while and start building more code using Observables and the various combinators (=operators?) then it will be a good opportunity to introduce it in more places as well. Until then, I'm hesitant to do that. Looking at the ngrx internal implementation, even though it's got a very small footprint it's not as easy to read and understand what is happening there. Redux, on the other hand, is very very simple and the way Dan presented it in the Egghead videos is just brilliant! Any developer can look at that and understand straight away what it does under the hood and how to reproduce it themselves. You can't say the same on the Rx implementation, thus creating another barrier for entry.

To summarise, with all the new concepts that our developers need to grasp when dealing with a modern web application (Angular 2, Typescript, transpilers, modules, bundling, debugging, material design, RxJS inside Angular, Redux as a concept, npm, etc.) I think an easier path to use Redux is the original implementation. That said, I can see a future where we can opt for another implementation, if the code using it is not tied in to the specific implementation (use simple dispatch/subscribe API).

Thanks for looking at the implementation and bringing up the question. Would love you hear your thoughts and any further research or choices you've made on this subject.
Cheers,
Ruby

**Comment by born2net** [comment]
thanks so much for the long explanation.... appreciate it...
I decided to take your advice and go with std Redux..

I will use your code as my examples while I learn Redux...
posted that earlier today... just wondering how well it will scale on a very large store.

. tx again, Sean.

**Comment by rubyboy** [comment]
No worries.

The link you provided doesn't work. Not sure what the question is, but if it's a concern about a large data set I don't think there should be an issue there. Remember that the state is just an encapsulating object with references. It doesn't actually contain all the data, just links to other objects. So the data that's "in it" will exist this way or another. The only overhead is the wrapper object (very small) and the fact we're creating it again and again with every action, which I think is worth the benefits, but it's something to verify when building a large app.

BTW, here's a recent blog post from Victor Savkin on this subject: [http://victorsavkin.com/post/137821436516/managing-state-in-angular-2-applications](http://victorsavkin.com/post/137821436516/managing-state-in-angular-2-applications)

**Comment by born2net** [comment]
Thanks again, ya I deleted the link since I got the answer, which supports what you said above. Another question I had is why you didn't use Immutable.js? I plan on including it and wanted your take on it... (would be awesome if it was part of this example for best practices ... ) 

regards

Sean.

**Comment by rubyboy** [comment]
I was actually contemplating about that for a while (still am :) ). I really don't like the fact that immutable.js introduces new syntax above pure JS for arrays and objects. I think it ties the code to this implementation and requires new developers to learn it. I'd much rather have it "just work" and complain if someone was trying to mutate an object and not required any additional overhead for both learning and in code.
In that sense I think [Seamless Immutable](https://github.com/rtfeldman/seamless-immutable) is a better option. But then I looked at some performance benchmarks and wasn't happy with what I saw.
I ended up deciding, for now, to enforce immutability just by convention and best practice, and see how that goes. Once we start writing a bit of code (in a month or two we'll have heaps :) ) we can decide on the best approach here.

I do agree that this is an interesting are to watch. And maybe I'm wrong and we do need something from day 1. There are a few other options for enforcing immutability, but the above two are the most popular to date (would love to hear about other good options to consider). Maybe in a few months a new one will emerge and will be even better than these.

Cheers,
Ruby

**Comment by rubyboy** [comment]
BTW, I'm more than happy to receive PRs for stuff you'd like to add. Even if it's not what we would choose it doesn't hurt to include it if that's a best practice (and I do agree that immutable.js is so popular it might just be...).

**Comment by born2net** [comment]
got it... tx again...

**Comment by born2net** [comment]
I am trying to get jspm to bundle in my project but I keep getting an error of:

```
err  Error: Unable to calculate canonical name to bundle file:///C:/msweb/studioEnterprise/src/App.js. Ensure that this module sits within the baseURL or a wildcard path config.
         at getCanonicalNamePlain (C:\msweb\studioEnterprise\node_modules\jspm\node_modules\systemjs-builder\lib\utils.js:203:13)
         at getCanonicalName (C:\msweb\studioEnterprise\node_modules\jspm\node_modules\systemjs-builder\lib\utils.js:130:19)
         at C:\msweb\studioEnterprise\node_modules\jspm\node_modules\systemjs-builder\lib\arithmetic.js:96:38
root|studioEnterprise>
```

have you seen this before?
I also noticed that you bundle App.ts and I was wondering why? since all my source code is already been transpiled, does System.js really need to know about TS? can't it just bundle only the .js files?

tx as always,

Sean.
